OpenStack: fetch volume path from metadata service

petrutlucian94 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Updates the OpenStack cloud provider to use the Nova metadata service as a fallback when retrieving mounted PV disk paths. Note that the Nova instance device metadata will contain the disk address and bus, which allows finding its path. 

This is needed as the *standard* mechanism of retrieving disk paths is not available when running k8s under OpenStack Hyper-V hosts.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #55312

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

petrutlucian94 committed:
openstack: fetch volume path from metadata service

The OpenStack cloud provider retrieves mounted Cinder volume paths
by looking in /dev/disk/by-id, expecting the disk serial IDs (e.g.
SCSI ID) to include the volume ID.

The issue is that not all hypervisors are able to expose this.
For example, Hyper-V will just preserve the original Cinder volume
lun SCSI ID (without setting the volume id). For this reason,
disk path lookups will fail.

In order to be able to leverage Hyper-V based OpenStack providers,
as a fallback, we're querying the metadata service, searching for
disk device metadata. Note that starting with Nova Queens, the Hyper-V
driver always provides disk address information through the instance
metadata.

FengyunPan commented:
/ok-to-test

FengyunPan commented:
/sig openstack

FengyunPan commented on pkg/cloudprovider/providers/openstack/openstack_volumes.go:
```diff
@@ -327,10 +334,70 @@ func (os *OpenStack) GetDevicePath(volumeID string) string {
+				return ""
+			}
+
+			if len(diskPaths) == 1 {
```
How about checking "0" and ">1"?
If "0", we can return "" soon.

petrutlucian94 replied:
Good point, I'm updating the commit.

petrutlucian94 replied:
Updated the commit, immediately returning if getting none or multiple disk paths.

While at it, I've also updated the log messages according to the k8s developer guide. Some of them were not formatted properly.

FengyunPan replied:
Have you updated the commit?
But I can't find the codes which check "0".

petrutlucian94 replied:
Yes, well, I don't explicitly check for '0': if there's exactly one disk path, I return it, otherwise I'm logging an error message and returning an empty string immediately: https://github.com/petrutlucian94/kubernetes/blob/fea1939cd7684169357fc13fd2ff4cd28e9d8eed/pkg/cloudprovider/providers/openstack/openstack_volumes.go#L375-L383

(in case the previous link becomes unavailable due to a new commit, here's the snippet: http://paste.openstack.org/raw/625887/)

FengyunPan replied:
All right, thank you.

FengyunPan commented:
/retest

dims commented:
/assign @anguslees

dims commented:
@petrutlucian94 @FengyunPan - Is this ready to merge?

petrutlucian94 commented:
@dims - I've addressed the comments and successfully tested the patch on a Hyper-V OpenStack environment. All CI tests are passing. I've double checked the code and IMO it's good to go.

Thank you guys for reviewing it.

dims commented:
/approve

FengyunPan commented:
Nice work.
/lgtm

petrutlucian94 commented:
/test pull-kubernetes-unit
/test pull-kubernetes-e2e-kops-aws
